feat: typescript 3.6 support

IgorMinar opened the pull request:
BREAKING CHANGE: typescript 3.4 and 3.5 are no longer supported, please update to typescript 3.6

Partly based on #32908 but instead of just making the code compile with 3.6, this PR actually upgrades the typescript version and drops support for tsc 3.4 and 3.5. Due to blocking issues, it might be a good idea to merge #32908 in first, and merge this PR once the blocking issues are resolved.

Currently blocked on:
- [x] typescript 3.6.4 release, which should fix https://github.com/microsoft/TypeScript/issues/33295 
  - (PR: https://github.com/microsoft/TypeScript/pull/33321, merged commit: https://github.com/microsoft/TypeScript/commit/46ccaa29d64dc7614de5e42a0ac20c6990f32210)

TODOs:
- [x] uncomment disabled tests from https://github.com/angular/angular/pull/32908/files#diff-264a68237aa09f1a8fb16d5b112053f7

Optional:
- [ ] tsickle 0.38 release with tsc 3.6 support
  - tracking issue: https://github.com/angular/tsickle/issues/1065

IgorMinar committed:
feat: typescript 3.6 support

BREAKING CHANGE: typescript 3.4 and 3.5 are no longer supported, please update to typescript 3.6

Fixes #32380

mary-poppins commented:
You can preview 55a03f4 at https://pr32946-55a03f4.ngbuilds.io/.

mary-poppins commented:
You can preview 10c0e38 at https://pr32946-10c0e38.ngbuilds.io/.

gkalpak commented:
@IgorMinar, shouldn't this PR be combined with/based on #32908?

gkalpak commented:
Also, this PR fixes #32380.

lppedd commented:
Hi! I've noticed the `package.json` uses `3.6.3`. However I encountered this issue when compiling with it in `--prod`, https://github.com/microsoft/TypeScript/issues/33295, and it is scheduled for `3.6.4`. I don't know if the other changes fixed it, just wanted to tell you.

mary-poppins commented:
You can preview 4852dcc at https://pr32946-4852dcc.ngbuilds.io/.

mary-poppins commented:
You can preview cb8d778 at https://pr32946-cb8d778.ngbuilds.io/.

IgorMinar commented:
@lppedd thanks for heads up. I just hit the issue and I'll escalate to the typescript team too cut the 3.6.4 release and publish it on npm.

@gkalpak I think we'll end up merging #32908 before this PR lands - I updated the PR description to clarify the order of events.

filipesilva commented:
A couple more fixes that were missing are in separate commits in https://github.com/angular/angular/pull/33027.

filipesilva commented on packages/compiler-cli/ngcc/src/host/commonjs_host.ts:
```diff
@@ -186,7 +186,7 @@ export class CommonJsReflectionHost extends Esm5ReflectionHost {
     } else {
       const moduleInfo =
-          ts.resolveModuleName(moduleName, containingFile.fileName, this.compilerHost);
+        ts.resolveModuleName(moduleName, containingFile.fileName, this.program.getCompilerOptions(), this.compilerHost);
```
`resolveModuleName` now takes the compiler options before the `ModuleResolutionHost`.

filipesilva commented on packages/compiler-cli/src/transformers/api.ts:
```diff
@@ -16,7 +16,7 @@ export const SOURCE = 'angular' as 'angular';
   messageText: string;
   position?: Position;
-  next?: DiagnosticMessageChain;
+  next?: DiagnosticMessageChain[];
```
`ts.DiagnosticMessageChain` is now a tree https://github.com/microsoft/TypeScript/commit/ba9d8e2e81736854734f07f4441100f29f603bcd

filipesilva commented on packages/core/test/application_ref_spec.ts:
```diff
@@ -375,7 +375,7 @@ class SomeComponent {
 
            const compilerFactory: CompilerFactory =
-               defaultPlatform.injector.get(CompilerFactory, null);
+               defaultPlatform.injector.get(CompilerFactory, null)!;
```
These could be null, but it seems like they weren't properly detected before. Since this is a spec and it was passing before, it seems ok to add a non-null assertion.

filipesilva commented:
@googlebot I consent.

mary-poppins commented:
You can preview fb72ebc at https://pr32946-fb72ebc.ngbuilds.io/.

IgorMinar commented:
rebased + updated to 3.6.4 that was just released!

mary-poppins commented:
You can preview 20359dd at https://pr32946-20359dd.ngbuilds.io/.

mary-poppins commented:
You can preview 3f547a6 at https://pr32946-3f547a6.ngbuilds.io/.

mary-poppins commented:
You can preview fd7d622 at https://pr32946-fd7d622.ngbuilds.io/.

mary-poppins commented:
You can preview 6bea270 at https://pr32946-6bea270.ngbuilds.io/.

mary-poppins commented:
You can preview 520b8cb at https://pr32946-520b8cb.ngbuilds.io/.

mary-poppins commented:
You can preview 7cc4c4a at https://pr32946-7cc4c4a.ngbuilds.io/.

mary-poppins commented:
You can preview 9a78135 at https://pr32946-9a78135.ngbuilds.io/.

mary-poppins commented:
You can preview 049f98b at https://pr32946-049f98b.ngbuilds.io/.

mary-poppins commented:
You can preview 4bc1075 at https://pr32946-4bc1075.ngbuilds.io/.

mary-poppins commented:
You can preview 85c560f at https://pr32946-85c560f.ngbuilds.io/.

mary-poppins commented:
You can preview bfa131f at https://pr32946-bfa131f.ngbuilds.io/.

mary-poppins commented:
You can preview 98db61e at https://pr32946-98db61e.ngbuilds.io/.

mary-poppins commented:
You can preview 64b33b1 at https://pr32946-64b33b1.ngbuilds.io/.

mary-poppins commented:
You can preview 6ff7ca1 at https://pr32946-6ff7ca1.ngbuilds.io/.

mary-poppins commented:
You can preview 0352229 at https://pr32946-0352229.ngbuilds.io/.

Commit:
fix(bazel): Remove angular devkit and restore ngc postinstall

This commit removes `@angular-devkit/build-angular` from package.json
for a project that opts into Bazel. This is because the package adds a
dependency on node-sass, which is rejected by Bazel due to its absense.

This commit also appends to `scripts.postinstall` if it already exists.
This is needed because `ng new` in CLI v9 now automatically adds a
postinstall step for `ngcc`.

mary-poppins commented:
You can preview ffdae82 at https://pr32946-ffdae82.ngbuilds.io/.

mary-poppins commented:
You can preview ab9700e at https://pr32946-ab9700e.ngbuilds.io/.

filipesilva commented:
@IgorMinar integration tests are showing this error:
```
ERROR in node_modules/@types/node/index.d.ts(73,11): error TS2300: Duplicate identifier 'IteratorResult'.
node_modules/typescript/lib/lib.es2015.iterable.d.ts(41,6): error TS2300: Duplicate identifier 'IteratorResult'.
```

This seems to be related to https://github.com/microsoft/TypeScript/issues/33344, where compatible `@types/node` versions need to be used. I think CLI needs to add a migration to update both TS and `@types/node` to compatible versions when going up to version 9. Meanwhile I'll changed versions here.

filipesilva committed:
test: use @types/node compatible with TS 3.6

filipesilva commented:
Note: this PR now needs to be rebase on master because of a conflict. I didn't rebase myself in order to maintain the current commit history, otherwise other people working on it would have to force pull.

mary-poppins commented:
You can preview c7c1ee5 at https://pr32946-c7c1ee5.ngbuilds.io/.

mary-poppins commented:
You can preview eb840af at https://pr32946-eb840af.ngbuilds.io/.

mary-poppins commented:
You can preview 0a6eba7 at https://pr32946-0a6eba7.ngbuilds.io/.

filipesilva commented on package.json:
```diff
@@ -126,7 +126,8 @@
   "// 3": "when updating @bazel/bazel version you also need to update the RBE settings in .bazelrc (see https://github.com/angular/angular/pull/27935)",
   "devDependencies": {
-    "@angular/cli": "^8.0.0-beta.15",
+    "@angular-devkit/build-angular": "^0.900.0-next.11",
```
`@angular-devkit/build-angular` has a peerdep on `@angular/compiler-cli`, which won't work very well here. That's why it's in all integration projects instead.

filipesilva replied:
https://github.com/angular/angular/pull/32957 updates all the integration tests to latest CLI btw.

mary-poppins commented:
You can preview ebb4214 at https://pr32946-ebb4214.ngbuilds.io/.

mary-poppins commented:
You can preview 6394f30 at https://pr32946-6394f30.ngbuilds.io/.

mary-poppins commented:
You can preview 72d9562 at https://pr32946-72d9562.ngbuilds.io/.

mary-poppins commented:
You can preview b530dc4 at https://pr32946-b530dc4.ngbuilds.io/.

IgorMinar commented:
merge-assistance: global approval

IgorMinar committed:
fixup! feat: typescript 3.6 support

mary-poppins commented:
You can preview 6737387 at https://pr32946-6737387.ngbuilds.io/.

mary-poppins commented:
You can preview 7f68455 at https://pr32946-7f68455.ngbuilds.io/.

IgorMinar committed:
fixup! feat: typescript 3.6 support

mary-poppins commented:
You can preview 4596f70 at https://pr32946-4596f70.ngbuilds.io/.

IgorMinar commented:
merge-assistance: g3 failures are flakes

petebacondarwin commented:
🎉
